packages/server-wallet/src/onchain-service/index.ts

```diff
@@ -0,0 +1,228 @@
+import {ContractArtifacts} from '@statechannels/nitro-protocol';
+import {providers, Wallet, Contract, BigNumber} from 'ethers';
+import PriorityQueue from 'p-queue';
+
+import {Wallet as ChannelWallet} from '../wallet';
+import {Address, Bytes32} from '../type-aliases';
+
+const BAD_NONCE = "the tx doesn't have the correct nonce";
+const INVALID_NONCE = 'Invalid nonce';
+const NO_TX_HASH = 'no transaction hash found in tx response';
+const UNDERPRICED_REPLACEMENT = 'replacement transaction underpriced';
+const DEFAULT_MAX_RETRIES = 5;
+const KNOWN_ERRORS = [BAD_NONCE, NO_TX_HASH, UNDERPRICED_REPLACEMENT, INVALID_NONCE];
+
+type OnchainServiceEnvironment = {
+  provider: string | providers.JsonRpcProvider;
+  wallet: string | Wallet; // Private key or wallet to send tx
+  transactionRetries?: number; // Maximum number of times a tx is retried
+};
+
+// This is used instead of the ethers `Transaction` because that type
+// requires the nonce and chain ID to be specified, when sometimes those
+// arguments are not known at the time of creating a transaction.
+type MinimalTransaction = {
+  to: Address;
+  value: string;
+  data: string;
+};
+
+type AssetHolderInformation = {
+  assetId: Address;
+  contractAddress: Address;
```

Comment:
```suggestion
  tokenAddress: Address;
  assetHolderAddress: Address;
```

ERC20 asset holders already stores the token information, so this type probably isn't even necessary: you should be able to just provide the asset holder address.

Reply:
The generic `AssetHolder` contract doesn't have a reference to this property, how would I access it? Definitely would love to drop the type.

Reply:
Looks like you will have to change the contract to make this an explicitly public property before I am able to access it this way.

Reply:
Ahhh. That's too bad -- duplicating the data is not the worst thing in the world, but it would sure be nice if we could just get the token address from the asset holder contract.